# Apprentissage supervisé - Classification

## 1    Préparation des données
### 1.1    Téléchargement de données

In [2]:
from sklearn.datasets import fetch_openml
import pandas as pd

In [3]:
bdd = fetch_openml(name = "mnist_784", version = 1)

In [4]:
bdd

{'data': array([[0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        ...,
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.],
        [0., 0., 0., ..., 0., 0., 0.]]),
 'target': array(['5', '0', '4', ..., '4', '5', '6'], dtype=object),
 'frame': None,
 'categories': {},
 'feature_names': ['pixel1',
  'pixel2',
  'pixel3',
  'pixel4',
  'pixel5',
  'pixel6',
  'pixel7',
  'pixel8',
  'pixel9',
  'pixel10',
  'pixel11',
  'pixel12',
  'pixel13',
  'pixel14',
  'pixel15',
  'pixel16',
  'pixel17',
  'pixel18',
  'pixel19',
  'pixel20',
  'pixel21',
  'pixel22',
  'pixel23',
  'pixel24',
  'pixel25',
  'pixel26',
  'pixel27',
  'pixel28',
  'pixel29',
  'pixel30',
  'pixel31',
  'pixel32',
  'pixel33',
  'pixel34',
  'pixel35',
  'pixel36',
  'pixel37',
  'pixel38',
  'pixel39',
  'pixel40',
  'pixel41',
  'pixel42',
  'pixel43',
  'pixel44',
  'pixel45',
  'pixel46',
  'pixel47',
  'pixel48',
  'pixe

### 1.2    Information sur les données

#### 1.2.1.  Le retour de la fonction "fetch_openml" est un dictionnaire (mnist). Affichez ces clés.

In [5]:
bdd.keys()

dict_keys(['data', 'target', 'frame', 'categories', 'feature_names', 'target_names', 'DESCR', 'details', 'url'])

#### 1.2.2.  En utilisant les clés du dictionnaire "mnist", affichez :

 •  La taille des données (le nombre des features et la taille de chaque feature).  Par convention, les données sont à stocker dans une variable "X".
 

In [6]:
bdd['data'].shape

(70000, 784)

In [8]:
X = bdd['data']
# Taille d'une feature :
X[0]

array([  0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,   0.,
         0.,   0.,   0.,   0.,   0.,   0.,   0.,   

In [ ]:
from math import sqrt
print(sqrt(X[0].shape[0]))

 •  La taille des labels=classes (le nombre des labels). Par convention, les classes sont souvent à stocker dans une variable "y"
 

In [ ]:
bdd['target'].size

In [ ]:
Y = bdd['target']
Y

 •  Les différentes classes de la base de données 

In [ ]:
import numpy as np

In [ ]:
np.unique(Y)

 •  Une description détaillée de la base de données.

In [ ]:
bdd

### 1.2.3. 

A travers cette question, vous avez dû conclure que la base de données MNIST contient 70000 images en niveau de gris, et chaque image est caractérisée par un features de taille 784=28*28.

•  Affichez l’image de la première instance de la base de données.

In [ ]:
test = X[0].reshape(28,28)
from matplotlib import pyplot as pp 
pp.imshow(test)

•  Affichez la classe de la première instance

In [ ]:
Y[0]

### 1.2.4.  Affichez le type des labels (du premier label par exemple)

In [ ]:
type(Y[0])

### 1.2.5. 
A  travers  la  question  précédente,  vous  avez  dû  remarquer  que  le  type  des  labels  estune chaîne de caractères.   Il est préférable,  dans les projets d’apprentissage automatique,d’utiliser  des  chiffres.   Par  conséquent,  appliquez  le  casting  sur  les  labels  pour  les  trans-former à des entiers.

In [ ]:
Y = Y.astype(int)
type(Y[0])

In [ ]:
Y

### 1.3.    Répartition des données

Contrairement  aux  précédents  projets,  les  données  dans  MNIST  sont  déjà  partitionnées  en base  d’apprentissage  et  en  base  de  test.    Les  60  000  premières  images  composeront  la  base d’apprentissage et le reste des images constituera la base de test;
Écrivez un code qui met en exergue cette répartition tout en stockant les données de test/d’apprentissage et les classes de test/d’apprentissage dans 4 variables.

In [ ]:
X_train = X[0:60000]
X_train.shape

In [ ]:
X_test = X[60000:]
X_test.shape

In [ ]:
Y_train = Y[0:60000]
Y_train.shape

In [ ]:
Y_test = Y[60000:]
Y_test.shape

## 2    Apprentissage d’un classifieur binaire

Comme indiqué dans les objectifs,  ce projet vise à classifier les chiffres.   Une des solutions est d’utiliser un classifieur binaire qui est apte d’identifier que l’image représente bien le chiffre que nous cherchons ou non. Par exemple, nous nous focaliserons sur la reconnaissance du chiffre 5.

### 2.1    Apprentissage des données

#### 2.1.1. 
En utilisant les labels d’apprentissage et de test, écrivez un code qui stocke dans deux variables distinctes (une première pour la base d’apprentissage et une autre pour la base de test) True si le classe de l’instance est 5 et False Sinon.  De ce fait, nous traitons une classification binaire.

In [ ]:
Y_train_bin = (Y_train == 5)
Y_train_bin

In [ ]:
Y_test_bin = (Y_test == 5)
Y_test_bin

#### 2.1.2. 
Pour l’apprentissage des données, nous allons utiliser le classifieur Stochastic Gradient Descent (SGD). Utiliser ce modèle pour la prédiction sur un chiffre (exemple 5).

In [ ]:
Y_train.shape

In [ ]:
from sklearn.linear_model import SGDClassifier

In [ ]:
clf = SGDClassifier()
clf.fit(X_train, Y_train_bin)

In [ ]:
Y_predit = clf.predict(X_train[[0]])
Y_predit

### 2.2    Évaluation du modèle d’apprentissage sur les données d’apprentissage

La mesure de performance d’une méthode de classification est souvent plus délicate qu’une méthode de régression. Cela est dû au nombre de mesures de performance existantes dans la littérature.

#### 2.2.1    Taux de classification.
##### 2.2.1.1
Écrivez un code qui répartit la base d’apprentissage en base d’apprentissage et en base de validation en utilisant la méthode 3-fold cross-validation.  Optez pour la valeur "accuracy"pour l’argument "scoring" pour afficher :

•  le taux de classification (accuracy) de chaque fold



In [ ]:
from sklearn.model_selection import cross_val_score

In [ ]:
cvs = cross_val_score(clf, X_train, Y_train_bin, cv=3, scoring='accuracy')

In [ ]:
cvs

•  la moyenne des taux de classification

In [ ]:
print("L'accuracy moyenne est de", cvs.mean())

##### 2.2.1.2.  Dans cette question, vous allez créer un classifieur simple qui classifie toutes les images de MNIST comme "non-5".  Créez une classe Never5Classifier qui hérite de la classe Base Estimator. la classe BaseEstimator existe dans le sous-module "base" du module "sklearn"

In [ ]:
from sklearn.base import BaseEstimator

class Never5Classifier (BaseEstimator):
    pass

2.1.2.3.  Dans la classe Never5Classifier, créez :

•  Une méthode fit qui prend en argument les données et les labels. Cette méthode ne va rien retourner et par conséquent va contenir que le mot-clé "pass".  Vous allez implémenter cette méthode car l’héritage de la classe BaseEstimator l’exige.

•  Une méthode predict qui prend en argument les données et retourne une structure dedonnées ayant la taille des données et qui contient que la valeur False (= non-5). Utilisezla fonction "zeros" du module "numpy" avec un "dtype=bool"

In [ ]:
class Never5Classifier (BaseEstimator):
    
    def fit (self, donnees, labels):
        pass
    
    def predict (self, donnees):
        resultat = np.zeros((len(donnees),1), dtype=bool)
        return (resultat)

#### 2.2.1.4.  Créez un objet de la classe Never5Classifier

In [ ]:
objet = Never5Classifier()

#### 2.2.1.5
Testez le classifieur en utilisant une validation croisée de type 3-fold cross-validation. Optezpour la valeur "accuracy" pour l’argument "scoring" pour afficher :  le taux de classification(accuracy) de chaque fold; La moyenne des taux de classification.

In [ ]:
cvs = cross_val_score(clf, X_train, Y_train_bin, cv=3, scoring='accuracy')
cvs

In [ ]:
cvs.mean()

#### 2.2.2    Matrice de confusion:

##### 2.2.2.1

In [ ]:
from sklearn.model_selection import cross_val_predict
Y_pred = cross_val_predict(clf, X_train, Y_train_bin, cv = 3)

##### 2.2.2.2

In [ ]:
from sklearn.metrics import confusion_matrix

In [ ]:
print("Confusion matrix\n")
print(confusion_matrix(Y_train_bin, Y_pred), "\n")

In [ ]:
TN, FP, FN, TP = confusion_matrix(Y_train_bin, Y_pred).ravel()

print("True Positives(TP) =", TP, "\n")
print("False Negatives(FP) =", FP, "\n")
print("False Positives(FP) =", FP, "\n")
print("True Negatives(TN) =", TN, "\n")

#### 2.2.3.  Précision et rappel:
En utilisant le modèle déjà construit du classifieur SGD, prédisez les classes des donnéesd’apprentissage.    Pour  ce  faire,  utilisez  la  fonction  "cross_val_predict"  du  sous-module"model_selection" du module "sklearn".  Optez pour une validation croisée de type 3-foldcross validation.

##### 2.2.3.1. Calculez la précision, le rappel et le score F1 du modèle d’apprentissage

In [2]:
from sklearn.metrics import precision_recall_fscore_support

ModuleNotFoundError: No module named 'sklearn'

In [ ]:
precision_recall_fscore_support(Y_train_bin, Y_pred, average='micro')

#(precision, rappel, score F1 et "support")

In [ ]:
from sklearn.metrics import precision_score
precision_score(Y_train_bin, Y_pred)

In [ ]:
from sklearn.metrics import recall_score
precision_score(Y_train_bin, Y_pred)

In [ ]:
from sklearn.metrics import f1_score
f1_score(Y_train_bin, Y_pred)

##### 2.2.3.2. 
Pour classifier les instances, le classifieur SGD calcule un score en se basant sur sa fonction de décision. Si le score est supérieur à un seuil, il affecte la classe positive à l’instance sinon il affecte la classe négative. Affichez les scores des différentes instances de la base d’apprentissage via la fonction "cross_val_predict". Optez pour une valeur de "decision_function" pour l’argument "method" et une validation croisée de type 3-fold cross-
validation.

In [ ]:
Y_score = cross_val_predict(clf, X_train, Y_train_bin, cv = 3, method = "decision_function")

In [ ]:
Y_score

##### 2.2.3.3. Calculez les précisions et les rappels de chaque instance en utilisant la fonction "precision_recall_curve"

In [ ]:
from sklearn.metrics import precision_recall_curve

In [ ]:
precision_recall_curve(Y_train_bin, Y_score)

In [ ]:
precision_recall_curve(Y_train_bin, Y_score)

In [ ]:
from sklearn.metrics import plot_precision_recall_curve
plot_precision_recall_curve(clf, X_train, Y_train_bin)

##### 2.2.3.4. Tracez la courbe des précisions/rappels. L’axe des abscisses doit contenir les rappels et l’axe des ordonnées doit contenir les précisions. Interprétez les résultats.

#### 2.2.4 Courbe ROC

##### 2.2.4.1. Calculez le taux de faux positifs (tfp), le taux de vrais positifs (tvp) et les seuils utilisés pour classifier les données d’apprentissage. Pour ce faire, utilisez la fonction "roc_curve" du sous-module "metrics" du module "sklearn"

In [ ]:
from sklearn.metrics import roc_curve

In [ ]:
courbe_roc = roc_curve(Y_train_bin, Y_score)

#La première valeur du tuple de courbe_roc est le tpr, la deuxième est le fpr, la troisième correspond aux seuils.
courbe_roc

##### 2.2.4.2. Tracez la courbe des ROC. L’axe des abscisses doit contenir les taux de faux positifs et l’axe des ordonnées doit contenir les taux de vrais positifs. Ajoutez des titres à ces deux axes et activer le mode "grid" avec la fonction grid de matplotlib

In [ ]:
from sklearn.metrics import plot_roc_curve
plot_roc_curve(clf, X_train, Y_train_bin)

##### 2.2.4.3. Calculez l’AUC (Area Under the Curve) du modèle d’apprentissage. Pour ce faire, utilisez la fonction "roc_auc_score" du sous-module "metrics" du module "sklearn".

In [ ]:
from sklearn.metrics import roc_auc_score

In [ ]:
roc_auc_score(Y_train_bin, Y_score)

In [ ]:
#Autre méthode : 
from sklearn.metrics import auc

auc(courbe_roc[0], courbe_roc[1])

## 3.    Apprentissage d’un classifieur multiclasse


Dans cette partie, nous allons apprendre un classifieur à classifier les 10 classes (les chiffres de 0 à 9) de la base de données MNIST.

### 3.1. Apprentissage des données

#### 3.1.1 En se basant sur le classifieur SGD, faites l’apprentissage du modèle en se basant sur toutes les instances (images) d’apprentissage. Ainsi le modèle va apprendre à partir de toutes les classes de la BD et non pas juste à partir de deux classes comme c’était le cas dans la partie II de TP.

In [ ]:
classifieur = SGDClassifier()
classifieur.fit(X_train, Y_train)

#### 3.1.2. Via le modèle d’apprentissage bâti, prédisez la classe de la première instance de la base de données (celle de la question 3). La prédiction va être, cette fois-ci, un chiffre de 0 à 9.

In [ ]:
Y_predit = classifieur.predict(X_train)
Y_predit[0]

Pour le coup la première valeur est bonne donc on est satisfaits mais il y a une part d'aléatoire donc ce n'est pas garanti que sur un autre apprentissage, le modèle ne donne pas plutôt 3.

In [ ]:
# Le score préférentiellement utilisé pour un SGD est l'accuracy :
classifieur.score(X_train, Y_train)

In [ ]:
# On peut aussi regarder le R2
from sklearn.metrics import r2_score
r2_score(Y_train, Y_predit)

#### 3.1.3. En utilisant l’objet qui instancie la classe SGDClassifier :
• Affichez, via la méthode "decision_function", les 10 scores de décision utilisés par la
méthode SGD pour classifier la première instance de la base de données.

In [ ]:
Y_score = cross_val_predict(classifieur, X_train, Y_train, cv = 3, method = "decision_function")

In [ ]:
Y_score.shape

In [ ]:
Y_score[0]

• Affichez, via l’attribut "classes_", les différentes classes utilisées par le classifieur.

In [3]:
classifieur.classes_

NameError: name 'classifieur' is not defined

In [ ]:
classifieur.predict(X_train)

In [ ]:
Y_score = classifieur.predict(X_train)
Y_score[0]

Qu’est ce que vous remarquez ?

En regardant le Y_score issu de la cross validation avec la méthode decision_function, on remarque que pour la valeur d'indice 5, le score de décision est celui qui est le plus proche de 0. C'est donc le label qui correspond à ce score qui aura été choisi par notre classifieur.

On remarque qu'un autre score est relativement proche de zéro également : celui pour le label 3.

C'est normal car le classifieur a tendance à hésiter entre 3 et 5 pour la détermination du label de la première instance de la base de données.

### 3.2 Évaluation du modèle d’apprentissage sur les données d’apprentissage

#### 3.2.1 Taux de classification

##### 3.2.1.1. Écrivez un code qui répartit la base d’apprentissage en base d’apprentissage et en base de validation en utilisant la méthode 3-fold cross-validation. Optez pour la valeur "accuracy" pour l’argument "scoring" pour afficher : le taux de classification (accuracy) de chaque fold; la moyenne des taux de classification

In [ ]:
cvs = cross_val_score(classifieur, X_train, Y_train, cv=3, scoring='accuracy')
cvs

In [ ]:
cvs.mean()

##### 3.2.1.2. Une des techniques d’amélioration des taux de classification est le "Scaling". Vérifiez ceci en appliquant une standardisation sur les données d’apprentissage avec la classe StandardScaler.

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
scaler = StandardScaler()

In [ ]:
scaler.fit(X_train, Y_train)

X_train
X_train_scaled = X_train.copy()
scaler.transform(X_train_scaled)
X_train_scaled

In [ ]:
classifieur.fit(X_train_scaled, Y_train)

##### 3.2.1.3. Évaluez le modèle sur les nouvelles données d’apprentissage avec une validation croisée de type 3-fold cross-validation. Les résultats sont-ils meilleurs que ceux avant le Scaling?

In [ ]:
cvs = cross_val_score(classifieur, X_train_scaled, Y_train, cv=3, scoring='accuracy')
cvs

In [ ]:
cvs.mean()

On remarque un meilleur résultat avec un standard scaling préalable mais la différence est minime et peut tenir de l'aléatoire.

#### 3.2.2 Matrice de Confusion

###### 3.2.2.1. En utilisant le modèle déjà construit du classifieur SGD, prédisez les classes des données d’apprentissage.

In [ ]:
Y_pred = classifieur.predict(X_train)
Y_pred

###### 3.2.2.2. Affichez la matrice de confusion du modèle d’apprentissage. Pour ce faire, utilisez la fonction "confusion_matrix" du sous-module "metrics" du module "sklearn". La taille de la matrice de confusion, cette fois-ci, est 10x10 car on traite un problème de classification de 10 classes. Interprétez les résultats.

In [ ]:
import seaborn as sns
matrice_conf = confusion_matrix(Y_train, Y_pred)

pp.figure(figsize=(7,7))
sns.heatmap(matrice_conf, annot=True, fmt=".0f", linewidths=.5, square = True, cmap = 'viridis');
pp.ylabel('Actual label');
pp.xlabel('Predicted label');
pp.title("Matrice de confusion", size = 15)
pp.show()

On note une belle diagonale bien remplie de valeurs plus élevées allant de la première à la dernière case de la matrice, cela correspond aux valeurs correctement prédites par notre modèle.

En effet, dans une matrice de confusion (qu'elle soit binaire ou multiclasse comme ici), les colonnes représentent les vraies valeurs et les lignes les valeurs prédites.

Tout ce qui se trouve à l'extérieur de la diagonale correspond donc à des objets mal classées par notre modèle.

**Cas notables :** on peut voir que dans 339 cas de 7 réels, notre modèle se sera trompé et aura décelé un 9 à tort. C'est la plus haute valeur de faux négatif pour notre tableau. Le 7 semble donc être le chiffre qui donne le plus de fil à retordre à notre modèle.

Par opposition, dans d'autres cas, il est rarissime que notre modèle se trompe. Par exemple il n'a confondu qu'une seule fois un 9 réel avec un 6.

In [ ]:
confusion_matrix(Y_train, Y_train)

Le cas plus haut est pour illustration, il montre une matrice "parfaite" sans erreur de valeur mal prédite.

## 4.   Version avec CNN (Convolutional Neural Networks)


In [ ]:
# Si on fait des bêtises sur notre modèle, on peut supprimer l'instance ainsi :
# del model

In [ ]:
import tensorflow as tf

In [4]:
tf.__version__

NameError: name 'tf' is not defined

**Normalisation des données :**

In [ ]:
X_train2 = tf.keras.utils.normalize(X_train, axis = 1)
X_train2

In [ ]:
X_test2 = tf.keras.utils.normalize(X_test, axis = 1)
X_test2

**Instanciation du modèle :**

In [ ]:
# On instancie notre objet modèle.
model = tf.keras.models.Sequential()

**Création des layers :**

Couche d'input :

In [ ]:
# Couche d'input :

# tf.keras.layers.Flatten() est un équivalent de np.reshape pour aplatir une matrice
model.add(tf.keras.layers.Flatten())

Couches intermédiaires :

In [ ]:
# Deux couches, 128 neurones chacune, une fonction d'activation relu (rectified linear unit) :

model.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))
model.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))

Couche d'output :

In [ ]:
# Il y a 10 neurones, nos 10 classes pour les nombres !

model.add(tf.keras.layers.Dense(10, activation = tf.nn.softmax)) #tf.nn.softmax : densité de probabilité

**Compilation du modèle :**

In [ ]:
# Compilation du modèle avec l'optimiseur adam:
model.compile(optimizer = "adam", # choix de l'optimiseur,
                     loss = "sparse_categorical_crossentropy", # choix de la fonction coût
                     metrics = ["accuracy"]) # choix du ou des métriques qu'on va vouloir suivre

**Entraînement du modèle :**

In [ ]:
model.fit(X_train2, Y_train, epochs = 3)

In [ ]:
val_loss, val_acc = model.evaluate(X_test2, Y_test)
print(val_loss)
print(val_acc)

On peut tester de créer un modèle avec un optimiseur SGD pour voir s'il donne de meilleurs résultats :

In [ ]:
#del model_sgd
model_sgd = tf.keras.models.Sequential()
model_sgd.add(tf.keras.layers.Flatten())
model_sgd.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))
model_sgd.add(tf.keras.layers.Dense(128, activation = tf.nn.relu))
model_sgd.add(tf.keras.layers.Dense(10, activation = tf.nn.softmax)) 
model_sgd.compile(optimizer = "SGD", # choix de l'optimiseur,
                  loss = "sparse_categorical_crossentropy", # choix de la fonction coût
                  metrics = ["accuracy"]) # choix du ou des métriques qu'on va vouloir suivre
model_sgd.fit(X_train2, Y_train, epochs = 3)

val_loss, val_acc = model_sgd.evaluate(X_test2, Y_test)
print(val_loss)
print(val_acc)

Si on se limite à 3 epochs, il vaut mieux privilégier l'optimiseur "adam", il donne une meilleure accuracy que l'optimiseur "SGD".

**Enregistrement du modèle :**

**Création d'un array de prédiction avec notre modèle :**

In [ ]:
predictions = model.predict([X_test2])

In [ ]:
type(predictions)

On peut s'amuser à regarder instance par instance de notre bdd pour voir si le modèle prédit bien.

In [ ]:
np.argmax(predictions[68])

Affichons la valeur réelle de cette instance de la bdd :

In [ ]:
pp.imshow(X_test2[68].reshape(28,28), cmap = pp.cm.binary)